### ISSUE TITLE
[ Proposal ] Support for dynamic loading of component configuration in Dapr Self-hosted mode

### ISSUE THREAD

**Issue by akhilac1** [dapr/dapr#4760]
/area runtime
/area docs

## Describe the proposal 

Prototype :  [ https://github.com/akhilac1/dapr_delayedComponentLoading](https://github.com/akhilac1/dapr/tree/dapr_delayedComponentLoading)

## What?
This proposal outlines the approach to enable dynamic loading of components using HTTP and gRPC API, when Dapr is running in  *self-hosted* mode. Dapr should be able to persist the configuration in a statestore to gurantee reloading of the components in case of crash of the Dapr process. This proposal focuses on bindings, but can be extended to all supported Dapr components like pub-sub / statestore etc. as a follow-up

## Why?
Cloud-based serverless offerings like Azure Logic Apps, abstract the underlying complexity and integration concerns, enabling seamless interaction with external systems. These solutions build and manage connectors to enable such abstraction. Dapr bindings also addresses the same problem space.  However, components such as bindings cannot be added to a running Dapr process [ in self-hosted mode] and hence user should decide upfront on the various components they wish to connect to. As a result the integration experience in not efficient.

## How?
In self-hosted mode Dapr should supports dynamic loading of configuration, pushed as JSON/YAML payload over HTTP /  gRPC API. Component creation follows the existing flow in Dapr runtime. An additional caveat is to fail the request to dynamically enable component loading if state store is not configured. This is mandatory as Dapr will not be restored to the desired state if there is a failure

Proposed flow for loading the configuration of components is illustrated as follows : 
![63236931-39fb-4915-9dd3-bdcd24f2f09a](https://user-images.githubusercontent.com/3939554/176890210-9a1f0d5c-8f52-45ca-8e5f-d9557656bfa7.jpg)

This can be achieved by enabling the following features in Dapr  for stand-alone mode:

> 1. GET API to get schema of connection, request and response metadata for supported components 
> 2. POST API to request dynamically load component[s] - *Since component loading is an asynchronous process, the API will respond successful loading request creation. User should query back using metadata API to validate if component is loaded.*
> 3. PATCH API to dynamically update loaded components
> 4. DELETE API to dynamically delete the loaded components
> 5. Persist component configuration in state-store to enable configuration recovery in case of failures
> 6. Improve logging to enable 

[Draft API Specification](https://github.com/akhilac1/wip_proposals/blob/main/dapr-metadata-api-spec.json)

 **Related Proposals :** 
- make metadata in DAPR API more portable https://github.com/dapr/dapr/issues/4603
- dapr/sig-api proposal: define a spec for metadata keys in Dapr API https://github.com/dapr/sig-api/issues/4
- Capabilities API for components https://github.com/dapr/dapr/issues/4621
- CRD generation + discovery https://github.com/dapr/cli/issues/992

**Comment by akhilac1** [comment]
@halspang - There is a WIP implementation of hot-reloading for k8s. Hence it is excluded from the scope for standalone.  
Agree with your comment on getting this to work for both K8S and standalone. There is additional design to be included for K8S - the configuration should be updated in K8S too, so that a restart triggered from K8S will not init the configuration to the default state.

Why can't we use restart for updating configuration vs. dynamic loading - In this scenario user will not know the components they intend to use before hand. User will continue to add the components on the go.

The need for persisting in statestore is to enable recovery when there is a failure during Dapr runtime. This behaviour is similar to configuration checkpointing every time it is updated.

**Comment by artursouza** [comment]
I agree that this proposal should not be standalone only and should also target K8s mode. It seems that for this feature to work, a state store component would need to be provided and have some sort of attribute in configuration to set a given state store as the "storage for dynamic components".

Other things to consider:
* The new API should only be accessible by admin/ops and should return 403 if invoked by the APP. It will require some auth token.
* The "static components" should not be modified via this API.
* Same as pointed by @halspang, where the requirements for the state store must be clear. The discovery part can be hard since query API is not present in all state stores and the consistency level might be an issue (i.e., the query results do not reflect the latest master copy of the records - it can happen in DynamoDB, por example).

**Comment by yaron2** [comment]
Why do we need a state store? On k8s, sidecars are already updated whenever components are added or removed, and of course no state store is needed since k8s is the state store and handles persistence, reconciliation etc.

On self hosted mode we need a "similar" mechanism that watches the filesystem and updates the sidecar when there's an update. Since components are loaded from the filesystem, that should be the source of truth just like the k8s API is on k8s.

Requiring a state store for self hosted mode, other than being redundant given alternative means to perform hot-reloading, will be introducing the first ever coupling of a building blocks to Dapr's core functionality. Instead of Dapr runtime functionality enabling the use of state stores, this will require a building block to enable a runtime operations feature.

Other than being an unsound architecture, this also introduces security risks (storing components might contain secrets, among others) and will baffle users: can we use encryption when storing components? Can we query them? etc.

**Comment by akhilac1** [comment]
Thanks for the review, feedback and follow-up conversations for helping me gain clarity @artursouza  @yaron2 @halspang 

To simplify checkpoint and restore and ensure that K8S and standalone behavior are aligned, dynamic loading of components will be via filesystem events - add file to load/update configuration of a component and delete file to unload a component. I will investigate the implementation feasibility for this design and update the details in the proposal.

**Comment by yaron2** [comment]
> Thanks for the review, feedback and follow-up conversations for helping me gain clarity @artursouza @yaron2 @halspang
> 
> To simplify checkpoint and restore and ensure that K8S and standalone behavior are aligned, dynamic loading of components will be via filesystem events - add file to load/update configuration of a component and delete file to unload a component. I will investigate the implementation feasibility for this design and update the details in the proposal.

Sounds good, thanks!

**Comment by akhilac1** [comment]
Based on a recent review with the team -
- Loading is not of real value unless it is backed by unloading to achieve dynamic configuration update. For example, a user might want to update the attribute of a component which combines the unloading old configuration and reloading the component with new configuration. This requires all the building blocks to support graceful shutdown/un-init functionality
- This cannot be a time boxed operation as shutdown would require that the component would need to finish what it is doing, release resources 
- Reloading the component would require that we bring statefulness to the building block by persisting the current configuration - for example - configuration build block is stateless today and we would need to change this to allow the component to persist the subscriptions/ids etc. so that they can be reassigned and component comes back to its original configured state before restart.

Owing to these considerations, the next step is to revisit this proposal with dynamic unloading also as a P0 requirement - for both standalone and K8S modes.
